### ISSUE TITLE
bug: sites breaking due to XHR - "Object state must be opened" error

### ISSUE THREAD

**Issue by wrongsahil** [requestly/requestly#1980]
Example Site: https://www.greenmatters.com/p/how-to-recycle-batteries

### Error
`Object state must be opened`
![CleanShot 2024-07-19 at 23 19 54](https://github.com/user-attachments/assets/ea8add1f-f293-4e33-a309-9e75ef84a729)

> Make sure to turn off your adblocker (if using any)

### Temporary Solution
You can add this website to blocklist so that Requestly would not have any effect on that website. Please refer the documentation [here](https://docs.requestly.com/general/blocklist)
> Note: None of the Requestly rules will work on the website

**Comment by wrongsahil** [comment]
## Backstory
Requestly overrides XMLHttpRequest to achieve Response interception and modification.
https://github.com/requestly/requestly/blob/01d183c977ced9f977de8ac4b6e277167e8b9123/browser-extension/mv3/src/page-scripts/ajaxRequestInterceptor/xhr.js#L245-L249

## Root Cause
Requestly overrides the XMLHttpRequest object but somehow some other script modifies it again but not completely (leaves `XMLHttpRequest.setRequestHeader` and constructor as it is). So due to this, when we call `setRequestHeader.apply(this.rqProxyXhr, arguments);`, `this.rqProxyXhr` existed and was never opened. (Opened by some other scripts `.open` method)

**Comment by wrongsahil** [comment]
[https://socialcanvas-cdn.kargo.com/js/rh.min.js](https://socialcanvas-cdn.kargo.com/js/rh.min.js)
This file seems to be the culprit

<img width="791" alt="CleanShot 2024-07-20 at 01 40 26@2x" src="https://github.com/user-attachments/assets/08cdd0c9-ab77-4a24-bc45-6884a30c73ab">

**Comment by wrongsahil** [comment]
## XHR Interceptor loaded by content script
```
var iframe = document.createElement('iframe');
iframe.src = 'https://example.com';
document.body.appendChild(iframe);
```


## XHR Interceptor not loaded by content script
```
var iframe = document.createElement('iframe');
document.body.appendChild(iframe);
```

```
var iframe = document.createElement('iframe');
var html = '<body>Foo</body>';
iframe.src = 'data:text/html;charset=utf-8,' + encodeURI(html);
document.body.appendChild(iframe);
```

**Comment by stephen-spiny** [comment]
+1 

I've only recently seen this issue start to pop up but seems to cause failures in JWPlayer and the Snowplow library. I think ultimately we'd prefer Requestly just work if no solution can be made to make both work. Kargo (culprit script source) can be a significant ad revenue driver but there have been 0 times that I can remember where I had to debug Kargo with Requestly on.

**Comment by wrongsahil** [comment]
@stephen-spiny It should be easy to fix all these sites by adding a try catch. But some Requestly rules (Request and Response) will not work completely on the given site which includes these scripts. It should work for now, but in longer run we need to somehow inject the Requestly Interceptor when they fail

**Comment by wrongsahil** [comment]
Hi @stephen-spiny , We've released a fix in the latest version, v24.7.29. Please follow [these steps](https://docs.requestly.com/guides/how-to-update-requestly-extension-manually) to update the extension manually. Let us know if this resolves the issue.

**Comment by wrongsahil** [comment]
Closing this issue for now. @stephen-spiny feel free to reopen incase you are facing this issue again.



Some downsides of this fix (On sites which has script which modifies XHR without chaining and breaks Requestly XHR interceptor)
- Response and Request body rules will not work.
- Delay Rules for AJAX requests will not work.
- Authorization headers won't be redirected in case of cross-origin Redirect rules
- `rq_request_initiator_origin()` won't work in headers rule
